### PR TITLE
fix: CometReader.loadVector should not overwrite dictionary ids

### LINKED ISSUE #475: CometReader.loadVector should not overwrite dictionary ids
### Describe the bug

Java Arrow provides `Data.importVector` API which is used to import Arrow array/schema through C Data interface. The caller needs to provide dictionary provider to the API and `Data.importVector` fills dictionary value vectors into the provider with dictionary ids are keys.

The dictionary ids are keys used to look up correct dictionary values for dictionary-encoded arrays. So it is obvious that dictionary id should be unique. Otherwise, dictionary arrays will use incorrect dictionary values and cause issues.

In Java Arrow API. one class `SchemaImport` is used to maintain current dictionary id during importing an array. But one design drawback of this `Data.importVector` API is, this `SchemaImport` is initiated internally in `Data.importVector`. So the uniqueness is only guaranteed for the array. For example, if the array is nested type, all dictionary-encoded arrays in the array will have unique dictionary ids.

But once you import another array by calling `Data.importVector` again, the dictionary id is reset. It cannot provide unique dictionary ids for all arrays you are importing.

Due to the above issue, `CometReader.loadVector` which calls `Data.importVector` to import arrays from native code, will overwrite dictionary ids of other arrays.

This is the cause of test failures of `CometTPCDSQuerySuite` in #437.

### Steps to reproduce

_No response_

### Expected behavior

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by viirya**
## Which issue does this PR close?

Closes #475.

### CODE REVIEW

**Comment by viirya** [comment]
cc @sunchao @andygrove

**Review by andygrove — APPROVED** [review]

File: common/src/main/scala/org/apache/arrow/c/CometSchemaImporter.scala [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+package org.apache.arrow.c
+
+import org.apache.arrow.memory.BufferAllocator
+import org.apache.arrow.vector.types.pojo.Field
+
+/**
+ * This is a simple wrapper around SchemaImporter to make it accessible from Java Arrow.
+ */
+class CometSchemaImporter(allocator: BufferAllocator) {
```
Comment (sunchao) [inline common/src/main/scala/org/apache/arrow/c/CometSchemaImporter.scala:28]: should we implement this in Java?

File: common/src/main/scala/org/apache/spark/sql/comet/util/Utils.scala [hunk 2]
```diff
@@ -264,4 +265,24 @@ object Utils {
...
+
+  /**
+   * Imports data from ArrowArray/ArrowSchema into a FieldVector. This is basically the same as
+   * Java Arrow `Data.importVector`. `Data.importVector` initiates `SchemaImporter` internally
+   * which is used to fill dictionary ids for dictionary encoded vectors. Every call to
+   * `importVector` will begin with dictionary ids starting from 0. So, separate calls to
+   * `importVector` will overwrite dictionary ids. To avoid this, we need to use the same
+   * `SchemaImporter` instance for all calls to `importVector`.
+   */
+  def importVector(
```
Comment (sunchao) [inline common/src/main/scala/org/apache/spark/sql/comet/util/Utils.scala:277]: ditto: may be it's better to put this in a separate class in Java other than spark.sql.comet.util?

Comment (viirya) [reply common/src/main/scala/org/apache/spark/sql/comet/util/Utils.scala:277]: Okay

**Comment by viirya** [comment]
`CometSchemaImporter` now is implemented in Java.

**Comment by viirya** [comment]
Merged. Thanks @andygrove @sunchao

**Pull request merged**